dvc/objects/transfer.py

```diff
@@ -60,23 +69,62 @@ def create_tasks(executor, amount):
                 yield rel_path_infos.pop(task), task.result()
 
 
-def _transfer_directory(odb, from_fs, from_info, jobs, no_progress_bar=False):
+def _get_tree(odb, *transfer_args, no_progress_bar=False, **transfer_kwargs):
     tree = Tree(None, None, None)
 
     with Tqdm(total=1, unit="Files", disable=no_progress_bar) as pbar:
-        for (
-            entry_info,
-            (entry_tmp_info, entry_hash),
-        ) in _transfer_directory_contents(odb, from_fs, from_info, jobs, pbar):
-            obj = HashFile(entry_tmp_info, odb.fs, entry_hash)
+        for entry_info, obj in _transfer_directory_contents(
+            odb, *transfer_args, pbar=pbar, **transfer_kwargs
+        ):
             tree.add(entry_info.parts, obj)
 
+    return tree
+
+
+def _transfer_directory(
+    odb, from_fs, from_info, jobs, update=None, no_progress_bar=False
+):
+    if update:
+        stream_cls = StreamFile
+    else:
+        stream_cls = TransferredStreamFile
+
+    tree = _get_tree(
+        odb,
+        from_fs,
+        from_info,
+        jobs,
+        no_progress_bar=no_progress_bar,
+        stream_cls=stream_cls,
+    )
     tree.digest()
+
+    if update:
+        entries = tree.as_list()
+        existing = odb.hashes_exist(entry[PARAM_CHECKSUM] for entry in entries)
+        tree.replace_contents(
+            _get_tree(
+                odb,
+                from_fs,
+                from_info,
+                jobs,
+                excluded_paths=frozenset(
+                    entry[tree.PARAM_RELPATH]
+                    for entry in entries
+                    if entry[PARAM_CHECKSUM] in existing
+                ),
+                no_progress_bar=no_progress_bar,
+                stream_cls=TransferredStreamFile,
+            ).as_dict()
```

Comment:
Looks really complicated

Reply:
Also, looks like we are effectively filtering out the already existing objects, but this is actually something that we can figure out during `save()`ing, as `odb.add` can figure it out on itsown. The only problem is that in order to know what to exclude, we need its hash_info, which will result in potentially unneeded uplading in case of TransferredStreamFile, hence you not using it in case of `update`.